src/macros.jl

```diff
@@ -811,11 +811,59 @@ esc_nonconstant(x::Number) = x
+        _error("Unrecognized keyword argument $kwarg")
+    end
+    if category == :Default
+        category = :Cont
+    end
+    if objective == 0 && isempty(inconstraints)
+        Variable(m, lowerbound, upperbound, category, basename, start)
+    else
+        Variable(m, lowerbound, upperbound, category, objective, inconstraints, coefficients, basename, start)
+    end
+end
+
+variabletype(m::Model, t::Union{Type{Val{:Cont}}, Type{Val{:Int}}, Type{Val{:Bin}}, Type{Val{:SemiCont}}, Type{Val{:SemiInt}}}) = variabletype(m)
+function constructvariable!(m::Model, t::Union{Type{Val{:Cont}}, Type{Val{:Int}}, Type{Val{:Bin}}, Type{Val{:SemiCont}}, Type{Val{:SemiInt}}}; category::Symbol=:Default, extra_kwargs...)
+    cat = t.parameters[1]
+    if category != :Default && category != cat
+        _error("A variable cannot be both of category $cat and $category. Please specify only one category.")
+    end
+    constructvariable!(m; category=cat, extra_kwargs...)
+end
+
+variabletype(m::Model, t::Symbol) = variabletype(m, Val{t})
+function constructvariable!(m::Model, t::Symbol; extra_kwargs...)
+    constructvariable!(m, Val{t}; extra_kwargs...)
+end
+
 const EMPTYSTRING = ""
 
 variable_error(args, str) = error("In @variable($(join(args,","))): ", str)
 
+# @variable(m, expr, extra...; kwargs...)
+# where `extra` is a list of extra positional arguments and `extra_kwargs` is a list of keyword arguments.
+# It creates a new variable (resp. a container of new variables) belonging to the model `m` using `constructvariable!` to create the variable (resp. each variable of the container).
+# The following modifications will be made to the arguments before being passed to `constructvariable!`
+# * The `expr` argument will not be passed but the expression will be parsed to determine the kind of container needed (if one is needed) and
+#   additional information that will be passed as keyword arguments. The latter can always be given directly as keyword arguments to the macro.
+# * The `SDP` and `Symmetric` positional arguments in `extra` will not be passed to `constructvariable!`. Instead,
+#    * the `Symmetric` argument will check that the container is symmetric and only allocate one variable for each pair of non-diagonal entries.
+#    * the `SDP` argument will do the same as `Symmetric` but in addition it will specify that the variables created belongs to the SDP cone in the `varCones` field of the model.
+#   Moreover, if a category is passed in `extra` not as a symbol (e.g. `Bin` instead of `:Bin`), it will be transformed to a symbol before being
+#   passed to `constructvariable!`.
+# * The keyword arguments start, objective, inconstraints, coefficients, basename, lowerbound, upperbound, category may not be passed as is to
+#   `constructvariable!` since they may be altered by the parsing of `expr` and we may need to pass it pointwise if it is a container since
+#   `constructvariable!` is called separately for each variable of the container.
```

Comment:
Could you give a trivial, e.g. ``@variable(m, x >= 0)`` and nontrivial example, e.g., ``@variable(m, x[1:N,1:N], SDP, Poly())`` of how the macro creates calls to ``constructvariable!``?